src/main/kotlin/com/terraformation/backend/device/db/TimeseriesStore.kt

```diff
@@ -238,6 +239,24 @@ class TimeseriesStore(private val clock: Clock, private val dslContext: DSLConte
     return fetchHistory(startTime, endTime, count, timeseriesIds)
   }
 
+  /**
+   * Returns the timestamps from the given list that already have values for the given timeseries.
+   * This is used to detect duplicate insertion attempts.
+   */
+  fun checkExistingValues(
+      timeseriesId: TimeseriesId,
+      timestamps: Collection<Instant>
+  ): Set<Instant> {
+    return with(TIMESERIES_VALUES) {
+      dslContext
+          .select(CREATED_TIME)
+          .from(TIMESERIES_VALUES)
+          .where(TIMESERIES_ID.eq(timeseriesId))
+          .and(CREATED_TIME.`in`(timestamps))
```

Comment:
does this get optimized internally because of the nature of timestamps/range?

Reply:
You mean on the database side? Not exactly. It's fast because there is a unique index on `(timeseries_id, created_time)` which means the database can run this query purely by scanning the index, no need to read the table data.

But that's also true of the duplicate key detection (it can detect duplicates using the index). What really makes this faster than the previous code is the fact that we're querying all the timestamps in a single query, rather than a separate query for each one. One database round-trip vs. 1000 of them, basically.